docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,895 @@
+```
+
+specifying `Mesh` in `spec.targetRef` will be optional (this probably will be most used case). Minimal policy, targeting
+`MeshHTTPRoute` will look like this:
+
+```yaml
+type: MeshTimeout
+name: timeouts-on-route
+mesh: default
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+To introduce this change we need to deprecate specifying `Mesh*Route` in `spec.targetRef`.
+
+#### Additional validation
+
+When referencing Mesh*Route we should validate policy configuration to make sure user is applying only configuration 
+that can be applied on route.
+
+### Affected policies
+
+At the moment we can only target `Mesh*Routes` in MeshTimeout policy. This MADR is based on MeshTimeout policy.
+Looking at what can be configured at Envoy route we can also apply this to our MeshRetry policy. Rate limit is also configurable on route,
+but we are configuring rate limit on inbound traffic, which does not apply to this MADR. `MeshLoadBalancingStrategy` also
+configures hashing per route and we should probably add support for routes in the future
+
+#### Producer/consumer model for targeting Mesh*Route in other policies
+
+For MeshRetry our producer/consumer makes sense since producer config is a suggestion from owner of the service. If consumer
+wants to override it because they don't want to retry, they just want to fail fast or they want to retry on bigger number of errors
+they can configure it by applying consumer policy.
+
+MeshRetry example:
+
+```yaml
+# Producer policy with retry on 500 HTTP error code
+apiVersion: kuma.io/v1alpha1
+kind: MeshRetry
+metadata:
+  name: producer-retry
+  namespace: backend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: backend-route # backend-route must be a producer route, for it to be synced to other zones and work
+      default:
+        http:
+          numRetries: 3
+          backOff:
+            baseInterval: 10ms
+            maxInterval: 1s
+          retryOn:
+            - "500"
+---
+# Consumer policy with override to retry on all 5xx errors
+apiVersion: kuma.io/v1alpha1
+kind: MeshRetry
+metadata:
+  name: consumer-retry
+  namespace: frontend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: backend-route
+        namespace: backend-ns
+      default:
+        http:
+          retryOn:
+            - "5xx"
+```
+
+For MeshLoadBalancingStrategy adding producer/consumer model can be problematic, for load balancing and hashing. Since this
+is configured on outbound it falls into producer/consumer policy model. But who should have more power on deciding hashing rules?
+I think it should be producer who is an owner of service who decides how to apply this configuration which is not compliant
+with our consumer/producer model.
```

Comment:
I don't see why the consumer can't control everything?

Reply:
Added example to explain this